File: code/ch02/2.6.4.maxSubArray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        maxSum = nums[0]
        minSum = sum = 0
        for i in range(n):
            sum += nums[i]
            maxSum = max(maxSum, sum - minSum)
            minSum = min(minSum, sum)

        return maxSum

File: code/ch02/2.6.4.maxSubArray.java
import java.lang.Math;

public class Solution {
  public int maxSubArray(int[] nums) {
    int n = nums.length;
    int maxSum = nums[0], minSum = 0, sum = 0;
    for (int i = 0; i < n; i++) {
      sum += nums[i];
      maxSum = Math.max(maxSum, sum - minSum);
      minSum = Math.min(minSum, sum);
    }
    return maxSum;
  }
}
